### ISSUE TITLE
RFE: default to dist-git style per-package directories in rpmbuild

### ISSUE THREAD

**Issue by pmatilai** [rpm-software-management/rpm#1171]
The traditional topdir/{SPECS,SOURCES,BUILD,...} setup is absolutely horrible even with small number of packages to deal with. In the meanwhile, all the distros have adopted a directory per package setup where all the content related to a package is in one place. That seems so obvious it's hard to imagine the rationale behind the original SPECS/SOURCES/etc split, but the world was different back then. At any rate, it's way past time rpm gets along with the times and starts shipping with sane packager defaults. Changing defaults is easy, the tricky part is figuring out the transition for old users.

**Comment by voxik** [comment]
I completely support this. However, not sure if mock should keep the original structure. @xsuchy, @praiskup, thoughts?

**Comment by praiskup** [comment]
No thoughts from me ATM.  I think mock will have to adapt, whatever change
will be done here (mock uses target-chroot rpmbuild).

**Comment by xsuchy** [comment]
+1 to the suggestion.

From Mock POV it is not necessary to have a long transition period. It can be even cut-off (since version x.y rpmbuild will use the new structure and not the old one). That is because Mock has to support very old target chroots. So we will have to have some detection of this functionality anyway.

The detection can be based on version:

```
rpm -q rpm
rpm --version #hmm bad thing that this string is localized

```
or some function or variable in rpmlib

**Comment by praiskup** [comment]
Yes, from mock perspective it shouldn't be huge problem (I guess we'll
stop depending on defaults, and pre-define some pattern by macros).  Still,
having a month or two in advance to have a chance to wrap new mock
release is worth it.  What is the ETA, and any more details how the new
pattern will look like?

**Comment by voxik** [comment]
My point was that I would prefer the current structure for the Mock, while the dist-git structure is much preferable for creating SRPMs

**Comment by Conan-Kudo** [comment]
This will be somewhat problematic for both CentOS and Mageia. Both distributions still use the current structure. I don't mind if the default is switched, provided we do some smart things to handle it and make a way to fallback to the old way easy enough.

**Comment by Conan-Kudo** [comment]
> Can the `sources` file be used to distinguish the directory layout? E.g. assume that if `sources` file is around, the dist-git directory layout is used and keep the traditional structure otherwise.

No, because not everyone does it that way. openSUSE and PLD don't _have_ a file, OpenMandriva uses `.abf.yml`, and so on.

**Comment by voxik** [comment]
Can there be some (configurable) file used to distinguish the directory layout?

**Comment by DemiMarie** [comment]
> This will be somewhat problematic for both CentOS and Mageia. Both distributions still use the current structure. I don't mind if the default is switched, provided we do some smart things to handle it and make a way to fallback to the old way easy enough.

Can they switch to the new layout?

**Comment by voxik** [comment]
> Can there be some (configurable) file used to distinguish the directory layout?

Actually, could there be at least some configuration knob to switch the default instead of changing multiple variables?

**Comment by Conan-Kudo** [comment]
> > This will be somewhat problematic for both CentOS and Mageia. Both distributions still use the current structure. I don't mind if the default is switched, provided we do some smart things to handle it and make a way to fallback to the old way easy enough.
> 
> Can they switch to the new layout?

Not easily. For CentOS, the usage of the current structure is enforced by the automation Red Hat uses to export RHEL sources to the CentOS Git server.

For Mageia, they inherited the structure from Mandriva's SVN configuration and reworking their Dist-SVN to support a flat layout would require changing almost everything around the infrastructure (the build system, the source control scripts, the package build helpers, etc.).

**Comment by praiskup** [comment]
> Can they switch to the new layout?

Like `$ rpmbuild --define 'new_fancy_layout 1'`?  Something that could be overridden in `~/.rpmmacros`?
